packages/fxa-react/components/Checkbox/index.tsx

```diff
@@ -0,0 +1,141 @@
+  useState,
+  useCallback,
+  ChangeEvent,
+} from 'react';
+import { v4 as uuidv4 } from 'uuid';
+import { ReactComponent as Checkmark } from './checkmark.svg';
+
+export type CheckboxProps = {
+  defaultChecked?: boolean;
+  disabled?: boolean;
+  id?: string;
+  label?: string;
+  onClick?: (event: MouseEvent<HTMLInputElement>) => void;
+  onBlur?: (event: FocusEvent<HTMLInputElement>) => void;
+  onFocus?: (event: FocusEvent<HTMLInputElement>) => void;
+  onChange?: (event: ChangeEvent<HTMLInputElement>) => void;
+  readerText?: string;
+};
+
+export const Checkbox = ({
+  defaultChecked,
+  disabled,
+  id,
+  label,
+  onClick,
+  onBlur,
+  onFocus,
+  onChange,
+  readerText,
+}: CheckboxProps) => {
+  const [focussed, setFocussed] = useState<boolean>(false);
+  const [checked, setChecked] = useState<boolean>(defaultChecked === true);
+  const [hovered, setHovered] = useState<boolean>(false);
+  const [checkboxId, setCheckboxId] = useState<string | undefined>(id);
+  const readerId = `${checkboxId}-sr`;
+
+  useEffect(() => {
+    !checkboxId && setCheckboxId(uuidv4());
+  }, [checkboxId, id]);
+
+  const checkboxFocus = useCallback(
+    (event: FocusEvent<HTMLInputElement>) => {
+      setFocussed(true);
+      onFocus && onFocus(event);
+    },
+    [focussed]
+  );
+
+  const checkboxBlur = useCallback(
+    (event: FocusEvent<HTMLInputElement>) => {
+      setFocussed(false);
+      onBlur && onBlur(event);
+    },
+    [focussed]
+  );
+
+  const checkboxChange = useCallback(
+    (event: ChangeEvent<HTMLInputElement>) => {
+      setChecked(event.target.checked);
+      onChange && onChange(event);
+    },
+    [checked]
+  );
+
+  return (
+    <label
+      className="flex items-start"
+      onMouseEnter={() => {
+        setHovered(true);
+      }}
+      onMouseLeave={() => {
+        setHovered(false);
+      }}
+      data-testid="checkbox-container"
+    >
+      <span className="relative">
+        <input
+          className="absolute inset-0 w-full h-full opacity-0 cursor-pointer"
+          type="checkbox"
+          id={checkboxId}
+          aria-describedby={readerText ? readerId : undefined}
+          onFocus={checkboxFocus}
+          onBlur={checkboxBlur}
+          onChange={checkboxChange}
+          data-testid="checkbox-input"
+          {...{
+            defaultChecked,
+            disabled,
+            onClick,
+          }}
+        />
+        <span
+          className={`block border-2 rounded w-5 h-5 transition-all duration-100 ease-in-out relative ${
+            disabled
+              ? 'border-grey-300 text-grey-300'
+              : `cursor-pointer ${
+                  checked
+                    ? `border-blue-500 text-blue-500 ${
+                        focussed ? 'shadow-input-blue-focus' : ''
+                      }`
+                    : `border-grey-400 text-grey-400 ${
+                        focussed ? 'shadow-input-grey-focus' : ''
+                      }`
+                }`
```

Comment:
Are `focussed` and `hovered` states used just for classes, specifically in the checkbox component? Can we just use Tailwind's `hover:whatever` etc. classes instead? Even `:hover:not(disabled)` as custom CSS seems a little easier to follow than rerendering the component on hover/focus change for a class update.

Reply:
So, in this specific case, the "focussed" state would only be applied if the preceding "checked" evaluated to true, and that only happens when the real input is checked (since this is a faux checkbox and cannot be checked nor focussed natively). But I think there are some other areas where this can be improved, so good catch. I will eliminate non-CSS state changes where possible.